internal/provisioner/kops_provisioner_cluster_installation_beta.go

```diff
@@ -0,0 +1,424 @@
+	if err != nil {
+		return errors.Wrapf(err, "failed to update cluster installation %s", clusterInstallation.ID)
+	}
+
+	logger.Info("Updated cluster installation")
+
+	return nil
+}
+
+// UpdateClusterInstallation updates the cluster installation spec to match the
+// installation specification.
+func (provisioner *kopsCIBeta) UpdateClusterInstallation(cluster *model.Cluster, installation *model.Installation, clusterInstallation *model.ClusterInstallation) error {
+	logger := provisioner.logger.WithFields(log.Fields{
+		"cluster":      clusterInstallation.ClusterID,
+		"installation": clusterInstallation.InstallationID,
+	})
+
+	configLocation, err := provisioner.getCachedKopsClusterKubecfg(cluster.ProvisionerMetadataKops.Name, logger)
+	if err != nil {
+		return errors.Wrap(err, "failed to get kops config from cache")
+	}
+	defer provisioner.invalidateCachedKopsClientOnError(err, cluster.ProvisionerMetadataKops.Name, logger)
+
+	k8sClient, err := k8s.NewFromFile(configLocation, logger)
+	if err != nil {
+		return errors.Wrap(err, "failed to create k8s client from file")
+	}
+
+	installationName, err := provisioner.prepareClusterInstallationEnv(clusterInstallation, k8sClient)
+	if err != nil {
+		return errors.Wrap(err, "failed to prepare cluster installation env")
+	}
+
+	ctx := context.TODO()
+
+	mattermost, err := k8sClient.MattermostClientsetV1Beta.MattermostV1beta1().Mattermosts(clusterInstallation.Namespace).Get(ctx, installationName, metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to get mattermost installation %s", clusterInstallation.ID)
+	}
+
+	logger.WithField("status", fmt.Sprintf("%+v", mattermost.Status)).Debug("Got mattermost installation")
+
+	mattermost.ObjectMeta.Labels = generateClusterInstallationResourceLabels(installation, clusterInstallation)
+
+	version := translateMattermostVersion(installation.Version)
+	if mattermost.Spec.Version == version {
+		logger.Debugf("Mattermost installation already on version %s", version)
+	} else {
+		logger.Debugf("Mattermost installation version updated from %s to %s", mattermost.Spec.Version, installation.Version)
+		mattermost.Spec.Version = version
+	}
+
+	if mattermost.Spec.Image == installation.Image {
+		logger.Debugf("Mattermost installation already on image %s", installation.Image)
+	} else {
+		logger.Debugf("Mattermost installation image updated from %s to %s", mattermost.Spec.Image, installation.Image)
+		mattermost.Spec.Image = installation.Image
+	}
+
+	// A few notes on installation sizing changes:
+	//  - Resizing currently ignores the installation scheduling algorithm.
+	//    There is no good interface to determine if the new installation
+	//    size will safely fit on the cluster. This could, in theory, be done
+	//    when the size request change comes in on the API, but would require
+	//    new scheduling logic. For now, take care when resizing.
+	//    TODO: address these issue.
```

Comment:
This doesn't need to be done for this PR, but would you mind opening a JIRA ticket for this TODO? This seems like a task we really want to do and don't want to just let it languish in a code comment, I think this is something we want to get put into the backlog so it can be prioritized.

Reply:
Sure, that is pretty old Todo, I have just copied it together with other logic from the ClusterInstallation part but I think it is still valid.
I will create an issue.